ui/appui-layout-react/src/appui-layout-react/state/internal/WidgetStateHelpers.ts

```diff
@@ -0,0 +1,236 @@
+    throw new UiError(category, "Floating widget does not exist");
+
+  state = produce(state, (draft) => {
+    delete draft.floatingWidgets.byId[id];
+    const idIndex = draft.floatingWidgets.allIds.indexOf(id);
+    draft.floatingWidgets.allIds.splice(idIndex, 1);
+  });
+  return removeWidgetState(state, id);
+}
+
+/** Removes floating widget from the UI and deletes the widget state.
+ * @internal
+ */
+export function removePopoutWidget(state: NineZoneState, id: PopoutWidgetState["id"]) {
+  if (!(id in state.popoutWidgets.byId))
+    throw new UiError(category, "Popout widget does not exist");
+
+  state = produce(state, (draft) => {
+    delete draft.popoutWidgets.byId[id];
+    const index = state.popoutWidgets.allIds.indexOf(id);
+    draft.popoutWidgets.allIds.splice(index, 1);
+  });
+  return removeWidgetState(state, id);
+}
+
+/** @internal */
+export function removePanelWidget(state: NineZoneState, id: WidgetState["id"], location?: PanelWidgetLocation): NineZoneState {
+  location = location || findPanelWidget(state, id);
+  if (!location)
+    throw new UiError(category, "Panel widget not found");
+
+  const panel = state.panels[location.side];
+  const widgets = [...panel.widgets];
+  widgets.splice(location.index, 1);
+  state = updatePanelState(state, panel.side, {
+    widgets,
+  });
+
+  const expandedWidget = widgets.find((widgetId) => {
+    return !state.widgets[widgetId].minimized;
+  });
+  if (!expandedWidget && widgets.length > 0) {
+    const firstWidgetId = widgets[0];
+    state = updateWidgetState(state, firstWidgetId, {
+      minimized: false,
+    });
+  }
+
+  return removeWidgetState(state, id);
+}
+
+function findPanelWidget(state: NineZoneState, id: WidgetState["id"]) {
+  const location = findWidget(state, id);
+  if (location && isPanelWidgetLocation(location))
+    return location;
+  return undefined;
+}
+
+/** @internal */
+export function assertWidgetState(state: NineZoneState, id: WidgetState["id"]) {
+  if (!(id in state.widgets))
+    throw new UiError(category, "Widget does not exist", undefined, () => ({ id }));
+}
+
+/** @internal */
+export function getWidgetState(state: NineZoneState, id: WidgetState["id"]) {
+  assertWidgetState(state, id);
+  return state.widgets[id];
+}
+
+/** @internal */
+export function setWidgetActiveTabId(state: NineZoneState, widgetId: WidgetState["id"], tabId: WidgetState["activeTabId"]): NineZoneState {
+  const widget = getWidgetState(state, widgetId);
+  if (!widget.tabs.includes(tabId))
+    throw new UiError(category, "Tab is not in a widget");
+
+  state = updateWidgetState(state, widgetId, {
+    activeTabId: tabId,
+  });
+
+  const floatingWidget = state.floatingWidgets.byId[widgetId];
+  if (floatingWidget) {
+    const activeTab = state.tabs[tabId];
+    const preferredFloatingWidgetSize = Rectangle.create(floatingWidget.bounds).getSize();
+    state = updateTabState(state, activeTab.id, {
+      preferredFloatingWidgetSize,
+    });
```

Comment:
Yes, `preferredFloatingWidgetSize` is used to create a new floating widget. This line is needed to maintain a widget size in case it was created from a previously active tab (another strategy would be to update `preferredFloatingWidgetSize` of all the tabs that are in a floating widget once added/resized, instead of an active one).